New keyshare protocol

sietseringers opened the pull request:
This provably secure keyshare protocol forces the user to compute and commit to its commitment before it gets access to that of the keyshare server, and similarly for the response.

There are the following differences with the old keyshare protocol:

- The [`ProofP`](https://github.com/privacybydesign/gabi/blob/72f2bc2b4d190cf14780a7d5d3d55e02802ef9c5/proofs.go#L362) instance produced at the end of the protocol by the keyshare server no longer contains the value `P mod N`, where `P = R_0^{keyshareserversecret}`, but only the challenge `c` and response `s`. This means that the ProofP no longer depends on the issuer public key; that is, a single ProofP can be used for all public keys. In addition, this means that a ProofP by itself is no longer linkable to the user.
- There is no longer a difference between issuance and disclosure: in both cases, a ProofP is sent by the user to the requestor alongside the zero-knowledge proofs.
- The zero-knowledge proofs as sent by the user to the requestor are always valid by themselves. That is, if the requestor just ignores the ProofP, then the zero-knowledge proofs are ordinary, valid zero-knowledge proofs. However, doing so ensures that the keyshare server's contributions to the zero-knowledge proofs are taken into account. This provides compatability with parties unaware of the keyshare server.

In this protocol, when the user requests the keyshare server's responses it must send a hash over its own commitments, including identifiers for the relevant issuer public keys. Such identifiers exist in [`irmago`](https://github.com/privacybydesign/irmago/blob/v0.10.0/identifiers.go#L48) but not here in `gabi`. Therefore, this PR uses generics in order to avoid referring to a specific type.

The fact that the user must send a hash over its commitments early in the protocol means that during one session, the user must compute its commitments for the zero-knowledge proofs twice: once to produce the hashed value to send to the keyshare server when requesting its commitments, and once after that when computing the challenge to use in the zero-knowledge proof. This PR uses the `Proof.Commit()` function for that, which is thus now invoked twice within a single session. Therefore, this PR also makes all `Commit()` functions idempotent, which additionally simplifies some state bookkeeping at some places in the code.

The use of generics makes it possible for the `KeyshareResponse` function, which contains the majority of the new code, to receive the relevant issuer public keys in the form of a parameter of type `map[T]*gabikeys.PublicKey`, where `T` is the generic type for the key identifier (i.e., in practice, `irma.PublicKeyIdentifier`). This seems to me a sensible type to transfer such information, but it deviates from the rest of `gabi`: other functions dealing with multiple proof (builders) take a `[]*gabikeys.PublicKey`), and then iterate index-wise. See e.g. [`ProofList.Verify()`](https://github.com/privacybydesign/gabi/blob/96cedcf/prooflist.go#L77). In the case of this PR I think the type `map[T]*gabikeys.PublicKey` makes more sense, since we must have such identifiers anyway as explained above, but we might also decide that consistency with the rest of `gabi` is more important and use `[]*gabikeys.PublicKey` instead.

sietseringers committed:
feat: initial support for new keyshare protocol

sietseringers committed:
refactor: remove keyshare PK commitment state from proof builders

sietseringers committed:
refactor: make CredentialBuilder.Commit idempotent by moving randomizer generation to constructor

sietseringers committed:
test: combined issuance and disclosure in keyshare response

sietseringers committed:
test: extend test coverage for keyshare response

sietseringers committed:
test: keyshare response for mixed keyshare servers

sietseringers committed:
test: refactor keyshare response subtests

sietseringers committed:
feat: update MergeProofP to new protocol version

sietseringers committed:
test: in case of issuance, check credential construction in new keyshare protocol

sietseringers committed:
feat: support revocation and range proofs in new keyshare protocol

sietseringers committed:
feat: make rangeproof commitment funcion idempotent

sietseringers committed:
test: revocation and range proofs in new keyshare protocol

sietseringers committed:
test: fix keyshare server tracking in keyshare response tests

sietseringers committed:
refactor: rename keyshare response functions

sietseringers committed:
docs: add comments to new keyshare code

sietseringers committed:
test: randomblind credentials in keyshare response

sietseringers committed:
test: pass keyshare server's ProofP to BuildDistributedProofList instead of merging it in manually

sietseringers committed:
test: signature sessions in keyshare response

sietseringers committed:
test: rearrange new test code

sietseringers committed:
feat: MergeProofP also sets challenge for new keyshare protocol

sietseringers committed:
docs: make gofmt happy

synaptic-cleft commented on builder.go:
```diff
@@ -96,16 +96,34 @@ func NewCredentialBuilder(pk *gabikeys.PublicKey, context, secret *big.Int, nonc
+		pk:           pk,
+		context:      context,
+		secret:       secret,
+		vPrime:       vPrime,
```
Reordering the parameters so that `vPrime` and `mUser` are named together, followed by their commitments would be more readable.

sietseringers replied:
For consistency with the struct definition of `CredentialBuilder`, I will name `vPrime` followed by its commitment, followed by `mUser` followed by its commitment.

synaptic-cleft commented on builder.go:
```diff
@@ -213,23 +231,19 @@ type CredentialBuilder struct {
 	mUserCommit map[int]*big.Int
 }
 
 func (b *CredentialBuilder) MergeProofPCommitment(commitment *ProofPCommitment) {
```
Since the merging functionality is removed, this method name should be updated.

synaptic-cleft commented on keyshare.go:
```diff
@@ -1,29 +1,145 @@
+	challengeContribs := make([]*big.Int, 0, len(res.ChallengeInput)*2)
+
+	// Assemble the input for the computation of h_W
+	for _, data := range res.ChallengeInput {
```
So the keyshare server check if `h_w` was constructed correctly is done in gabi. However, the assembly of `h_w` will be done in irmago. Can the assembly also be moved to gabi? I think this makes sense since the order of appending values matters.

sietseringers replied:
What do you mean with "assembly"? The keyshare server (whose code this is) will receive the input parameters to `h_W = H(...)` from the irmaclient, so when when it gets them it can just forward it to this function. Or, do you mean the assembly of these parameters in the irmaclient code?

synaptic-cleft replied:
Yep, I was thinking of generating/assembling `h_w` in the irmaclient code in irmago.

sietseringers replied:
Good point, done.

synaptic-cleft replied:
My underlying point here is that I think this should be the same code since `h_w` is the same. That's why I thought it would be nice to have both the client and the server make use of the same function. Both `h_w` assemblies are now in `keyshare.go`. However, the code is still split and not very obvious the same. I guess you did not put the whole hashing functionality in one function because of linked surrounding code (e.g. `challengeContribs` needs the same for loops as the hashing etc).

It's still hard to read for me at least.
In `KeyshareResponse` the hash is introduced as `hashContribs := make([]KeyshareUserChallengeInput[T], 0, len(res.UserChallengeInput))` whereas in `KeyshareUserCommitmentRequest` it's `var hashInput []KeyshareUserChallengeInput[T]`. It would be nice if at the naming would be consistent.

sietseringers replied:
It's not the same code, because it can't be. The client computes the input parameters of `h_W` by invoking `Commit()` on its builders, which the keyshare server cannot do because it doesn't (and shouldn't) have those builders. Instead, the keyshare server directly takes whatever the client claims were its input parameters to `h_W = H(...)`, as sent by the client when it requests the proof response from the keyshare server.

So the client and keyshare server can't use the same code to compute `h_W` because of their different roles and knowledge. This is not new in `gabi`, a similar thing happens with the commitment in creating or verifying a zero-knowledge proof:
* The client computes the commitment as (schematically) `commitment := base^randomizer`;
* When verifying the proof, the verifier recomputes the commitment without having access to the `randomizer` value (which is private to the client), but instead from the proof that it receives from the client.

> In `KeyshareResponse` the hash is introduced as `hashContribs := make(...)` whereas in `KeyshareUserCommitmentRequest` it's `var hashInput []KeyshareUserChallengeInput[T]`. It would be nice if at the naming would be consistent.

I now use the `make(...)` syntax in `KeyshareUserCommitmentRequest` as well (apart from consistency that also avoids potential slice growing by creating the slice with the appropriate capacity). As to naming, per your other comment I have removed the `hashInput` variable from `KeyshareResponse` because it's not necessary there, so that inconsistency is now gone as well 😅

synaptic-cleft commented on keyshare.go:
```diff
@@ -1,29 +1,145 @@
+	}
+
+	// Check that h_W sent in the commitment request equals the hash over the expected values
+	b, err := KeyshareUserCommitmentsHash(hashContribs)
```
After reading this several times I figured `b` stands for the byte array. Would you mind to rename this to something more outspoken, e.g. `recalculatedHash`?

synaptic-cleft commented on keyshare.go:
```diff
@@ -1,29 +1,145 @@
+	challenge := createChallenge(res.Context, res.Nonce, challengeContribs, res.IsSignatureSession)
+
+	// Compute our response and return the total response
+	ourResponse := new(big.Int).Add(randomizer, new(big.Int).Mul(challenge, secret))
```
Rename `ourResponse` -> `keyshareResponse`

sietseringers replied:
I have replaced all occurences of `our` with `kss` (with the extra `s`, KeyShareServer, for consistency with other variable names elsewhere).

synaptic-cleft commented on proofs.go:
```diff
@@ -166,7 +171,12 @@ type ProofD struct {
 // MergeProofP merges a ProofP into the ProofD.
 func (p *ProofD) MergeProofP(proofP *ProofP, _ *gabikeys.PublicKey) {
-	p.SecretKeyResponse().Add(p.SecretKeyResponse(), proofP.SResponse)
+	if proofP.P == nil { // new protocol version
```
You replaced the `SecretKeyResponse()` method with usingn the `AResponses[0]` directly. Is this just for better readability? Have you considered removing the `SecretKeyResponse()` method in its entirety?

sietseringers replied:
Yes, for better readability; it seemd just simpler to refer to `p.AResponses[0]` directly.

As to removing it `SecretKeyResponse()` altogether, that's not possible; that method is part of the `Proof` interface, and [here](https://github.com/privacybydesign/gabi/blob/new-keyshare-protocol/prooflist.go#L110) that method is invoked on a `Proof` instance without knowing the actual type.

synaptic-cleft commented on proofs.go:
```diff
@@ -166,7 +171,12 @@ type ProofD struct {
 // MergeProofP merges a ProofP into the ProofD.
 func (p *ProofD) MergeProofP(proofP *ProofP, _ *gabikeys.PublicKey) {
-	p.SecretKeyResponse().Add(p.SecretKeyResponse(), proofP.SResponse)
+	if proofP.P == nil { // new protocol version
```
You differentiate here between the old and the new protocol which is fine. However, some sort of protocol versioning would be nice. If we get yet another new protocol version, this new and old will be confusing otherwise.

sietseringers replied:
I agree that this approach for dealing with different protocol versions doesn't scale, but if we want to have access to that version number in this particular method, that would mean that the version number would have to be either a new field on `ProofD`, or a parameter to `MergeProofP`. Neither would be pretty.

Additionally, of course I might be completely wrong but I don't reallly expect further versions of the protocol at all. The old one was the very first one, since 2016 or so.

For now I would prefer to keep this the way it is, and instead deal with version numbers if and when yet another protocol version would become necessary.

synaptic-cleft commented on keyshare.go:
```diff
@@ -1,29 +1,145 @@
+		// server does not participate for this value and commitment.
+		KeyID *T `json:"key,omitempty"`
+
+		Value      *big.Int `json:"val"`
```
If I read this correctly `Value` is what is `P` in the paper and `Commitment` is `W_u`, the user commitment. It would be nice if the naming would be more consistent between the theory (paper) and the code. I'm just not sure how `P` could be called if not just `P`. Do you have an idea for a better word than `Value`?

sietseringers replied:
For disclosure sessions it's `A' = AS^r`, for issuance it's `U`. So I chose not to call it `P` because in practice (i.e., here in this code), we always deal with more complicated things than just the `P` from the paper. So we might consider it "the thing whose exponents we want to prove knowledge of", or something, although that still cutting some corners. And it doesn't make for a pretty variable name.

So, no, I don't know a better word than `Value` for now 😅

synaptic-cleft commented on proofs.go:
```diff
@@ -44,11 +44,16 @@ type ProofU struct {
-	)
-	p.SResponse.Add(p.SResponse, proofP.SResponse)
+	if proofP.P == nil { // new keyshare protocol version
+		p.C.Set(proofP.C)
```
Again, if I read this correctly, the old version uses `KeyshareResponse` to create the proof and then both `P` and `C` get set directly. Now here, `P` is not set and so `C` is not set. This is why you're still setting it here. Is this correct?

synaptic-cleft replied:
And would it then not make more sense to deprecate the MergeProofP methods and move them to legacy?

sietseringers replied:
> Now here, `P` is not set and so `C` is not set. This is why you're still setting it here. Is this correct?

No, generally the `ProofU.C` will always be set; the irmaclient will always compute the challenge and include that in the proof it sends to the requestor. Here we overwrite it in case of the new protocol with `proofP.C`, since `proofP.C` comes from the verified keyshare server JWT, and since the keyshare server has enforced in the keyshare protocol that that challenge has been honestly constructed. So either the proof must verify against the keyshare server's `C` or we must consider it invalid. The simplest way to get that behaviour is to just set `p.C.Set(proofP.C)`.

You might say that since this code always overwrites the `C` that the client sends, the client might skip sending it entirely. But one of the nice things about this protocol (in my opinion) is that if a hypothetical non-IRMA Idemix implementation would want to verify the proof, then it can just ignore the keyshare server JWT entirely (since it will probably be unaware of the IRMA keyshare server protocol) and still validate the proof, although it will of course miss the security guarantees that the keyshare server gives. That would break if the client would not send its `C`.

> And would it then not make more sense to deprecate the MergeProofP methods and move them to legacy?

No, this method is used in both protocol versions.

synaptic-cleft commented on gabi_test.go:
```diff
@@ -592,26 +596,30 @@ func TestGenerateKeyPair(t *testing.T) {
 }
 
-func createCredential(t *testing.T, context, secret *big.Int, issuer *Issuer) *Credential {
+func createKeyshareCredential(t *testing.T, context, secret, keyshareP *big.Int, attrs []*big.Int, issuer *Issuer) *Credential {
```
Would be nice to keep the type for `secret *big.Int`

sietseringers replied:
Why? Its type is enforced by the type checker to be `*big.Int` by the `*big.Int` after `keyshareP`.

synaptic-cleft replied:
Oh darn, I did not read this correctly. You're right of course!

synaptic-cleft commented on gabi_test.go:
```diff
@@ -1150,6 +1158,384 @@ func TestConstructCredentialNonZeroRandomBlindAttributes(t *testing.T) {
+
+var context = bigOne
+
+func newCredbuilder(t *testing.T, pk *gabikeys.PublicKey, secret, ourP *big.Int, randomblind []int) *CredentialBuilder {
```
`newCredbuilder` -> `newCredBuilder` for consistency?

synaptic-cleft replied:
and `secret` -> `secret *big.Int`?

synaptic-cleft commented on gabi_test.go:
```diff
@@ -1150,6 +1158,384 @@ func TestConstructCredentialNonZeroRandomBlindAttributes(t *testing.T) {
 
+func TestKeyshareResponse(t *testing.T) {
+	// Setup secrets and randomizer
+	ourSecret, err := GenerateSecretAttribute()
```
I find `our` somewhat confusing. I would rather call it `ksSecret`.

synaptic-cleft commented on gabi_test.go:
```diff
@@ -1150,6 +1158,384 @@ func TestConstructCredentialNonZeroRandomBlindAttributes(t *testing.T) {
+	return disclosureBuilder
+}
+
+func ourP(ourSecret *big.Int, pk *gabikeys.PublicKey) *big.Int {
```
as above, I'd replace `our` with `ks`

synaptic-cleft replied:
Actually,  I'd do that for all `our`s. So I won't mention it again :)

synaptic-cleft commented on gabi_test.go:
```diff
@@ -1150,6 +1158,384 @@ func TestConstructCredentialNonZeroRandomBlindAttributes(t *testing.T) {
+	// secret key responses.
+	keyshareSecretRandomizerLength = gabikeys.DefaultSystemParameters[1024].Lstatzk
+
+	// Prepare some vars used throughout.
```
used throughout... the protocol?

synaptic-cleft commented on gabi_test.go:
```diff
@@ -1150,6 +1158,384 @@ func TestConstructCredentialNonZeroRandomBlindAttributes(t *testing.T) {
+		Nonce:              nonce,
+		UserResponse:       userResponse,
+		IsSignatureSession: signature,
+		ChallengeInput:     hashInput,
```
`HashInput` might be a better name than `ChallengeInput`

sietseringers replied:
I agree that `ChallengeInput` isn't a very good name, but I don't find `HashInput` very clarifying, because there is lots of hashing going on in all of Idemix.

How about `UserChallengeInput`? That makes it more clear that these parameters are the user's contribution to the input parameters of the challenge `c = H(...)`.

synaptic-cleft commented on gabi_test.go:
```diff
@@ -1150,6 +1158,384 @@ func TestConstructCredentialNonZeroRandomBlindAttributes(t *testing.T) {
+	return new(big.Int).Exp(pk.R[0], ourSecret, pk.N)
+}
+
+func TestKeyshareResponseSingleBase(t *testing.T) {
```
What's the added value of this test when there's already the other flexible one? I compared the logic but cannot see what's tested more here what could not already be tested in the other one.

sietseringers replied:
Yes, everything that this test covers is also covered in the other more flexible one. But the other one always uses the full machinery of disclosure and/or credential builders, while this one purely tests the new keysharing for just a single secret and base. So if this test would fail, then you can be sure that the problem is somewhere in the new keyshare code, while if the other one fails, then you can't (without further investigation) be sure if the error is there, or in the code where the keysharing interacts with proof builders.

If you feel that that benefit is too little then I would be fine with removing this test entirely.

synaptic-cleft commented on gabi_test.go:
```diff
@@ -1150,6 +1158,384 @@ func TestConstructCredentialNonZeroRandomBlindAttributes(t *testing.T) {
+		"DisclosureRevocationRangeProof": {builders: ProofBuilderList{
+			newDisclosureBuilder(t, testPrivK, testPubK, userSecret, ourP(ourSecret, testPubK), true, rangeStatements),
+		}},
+		"DoubleDisclosureSameKeys": {builders: ProofBuilderList{
```
Just out of curiosity (and lack of understanding): Do you have an example when such a double disclosure with the same keys would be used?

sietseringers replied:
For example, disclosing `pbdf.gemeente.personalData.fullName` and `pbdf.gemeente.address.municipality` in a single disclosure session. That would result in two disclosure builders being used that use the same public key, as is tested here. Not sure if there is currently an actual use case somewhere doing that, but regardless we certainly need to support it.

sietseringers committed:
refactor: reorder fields in returned CredentialBuilder for legibility

sietseringers committed:
refactor: rename MergeProofPCommitment to SetProofPCommitment

sietseringers committed:
refactor: give recalculated hash parameter a more descriptive name

sietseringers committed:
test: refactor: rename keyshare variables in keyshare test code

sietseringers committed:
test: refactor: rename newCredbuilder to newCredBuilder

sietseringers committed:
test: docs: clarify comment

sietseringers committed:
Merge branch 'master' into new-keyshare-protocol

sietseringers committed:
fix: total keyshare secret too large for 1024 bit keys

sietseringers committed:
docs: document KeyshareChallengeInput.Value and .Commitment

sietseringers committed:
refactor: rename KeyshareChallengeInput to KeyshareUserChallengeInput

sietseringers committed:
feat: add functions to be used by irmaclient to compute keyshare protocol messages

This adds the functions KeyshareUserCommitmentRequest() and KeyshareUserResponseRequest(), which can be used by the irmaclient to compute its first and second message respectively to the keyshare server in the new keyshare protocol. These functions abstract the cryptographic details of the protocol away from the irmaclient, and accept parameters that the irmaclient will mostly already have. They contain code that was previously in the testNewKeyshareResponse() test helper function, which has therefore become shorter and closer to the actual future implementation in the irmaclient.

synaptic-cleft commented on keyshare.go:
```diff
@@ -1,12 +1,208 @@
+
+	// Assemble the input for the computation of h_W
+	for _, data := range res.UserChallengeInput {
+		hashContribs = append(hashContribs, data)
```
Couldn't you just use `res.UserChallengeInput` instead of appending to the new `hashContribs` for the `h_w` input?

sietseringers replied:
Yes indeed, thanks for that 😅 I think `hashContribs` is a leftover from an early version of this branch in which it wasn't an exact copy of `res.UserChallengeInput`. But it is now, so I've removed it as you suggested.

synaptic-cleft commented on keyshare.go:
```diff
@@ -1,12 +1,208 @@
+func KeyshareResponse[T comparable](
+	secret *big.Int,
+	randomizer *big.Int,
+	req KeyshareCommitmentRequest,
```
Something is still a bit off in the naming here. Both `req` and `res` are requests, one for commitment and one for the response phase. So the names `req` and `res` are incoherent

sietseringers committed:
refactor: remove superfluous var in KeyshareResponse()

sietseringers committed:
refactor: rename req and res parameters in KeyshareResponse()

sietseringers committed:
feat: declare hashInput slice with appropriate capacity in KeyshareUserCommitmentRequest()

ivard reviewed (approved):
Confirmed that two internal developers approved.
